feat: add .NET extensions (#5807)
canonical/snapcraft: docs/reference/extensions/dotnet-extensions.rst

Reviewer: This is a larger ask, but could you wrap these lines at 88 chars?

Recently we disabled line lengths as a blocking lint check, but keeping to them for the time being is a nice courtesy to contributors who read on smaller displays and the CLI.

.. _reference-dotnet-extensions:

.NET extensions
===============

The .NET extensions provide support for building and running .NET application snaps. They are versioned extensions,
which means that you can specify which extension you want to use based on the .NET major version your application
targets. The available extensions are:

* ``dotnet8``: For applications targeting .NET 8 (``net8.0``)
* ``dotnet9``: For applications targeting .NET 9 (``net9.0``)
* ``dotnet10``: For applications targeting .NET 10 (``net10.0``)

The .NET extensions are compatible with the core24 base.

Included parts
--------------

The .NET extensions add the following parts to the project file:

* ``prereqs``: A part that installs the prerequisite packages for running .NET applications. The installed packages are:

>    * ``libicu74`` (for the core24 base)

Reviewer 2: ```suggestion
   * ``libicu74``
```

Since we currently only support core24

   * ``libssl3t64``
   * ``libunwind8``
   * ``liblttng-ust1t64``
   * ``libbrotli1``

* ``launcher``: A part that sets up the snap launcher to run the .NET application. It verifies whether the necessary
  .NET runtime is available through a content interface connection and launches the application using the ``dotnet`` command.

.. collapse:: Included parts

    .. code-block:: yaml
        :caption: snapcraft.yaml

        launcher:
            plugin: dump
            source: /snap/snapcraft/x4/share/snapcraft/extensions/dotnet

>             override-build: |2
>             mkdir -p $CRAFT_PART_INSTALL/bin/command-chain
>             cp launcher.sh $CRAFT_PART_INSTALL/bin/command-chain

Reviewer 2: ```suggestion
            override-build: |
                mkdir -p $CRAFT_PART_INSTALL/bin/command-chain
                cp launcher.sh $CRAFT_PART_INSTALL/bin/command-chain
```

.. _reference-dotnet-extensions:

.NET extensions
===============

> The .NET extensions provide support for building and running .NET application snaps. They are versioned extensions,

Reviewer: ```suggestion
The .NET extensions help package apps made with the .NET framework. They are versioned extensions,
```

A nitpick, but I think it's a bit more straightforward this way.

which means that you can specify which extension you want to use based on the .NET major version your application
targets. The available extensions are:

* ``dotnet8``: For applications targeting .NET 8 (``net8.0``)
* ``dotnet9``: For applications targeting .NET 9 (``net9.0``)
* ``dotnet10``: For applications targeting .NET 10 (``net10.0``)

The .NET extensions are compatible with the core24 base.

Included parts
--------------

The .NET extensions add the following parts to the project file:

* ``prereqs``: A part that installs the prerequisite packages for running .NET applications. The installed packages are:

>    * ``libicu74``
>    * ``libssl3t64``
>    * ``libunwind8``
>    * ``liblttng-ust1t64``
>    * ``libbrotli1``

Reviewer: ```suggestion
  * ``libicu74``
  * ``libssl3t64``
  * ``libunwind8``
  * ``liblttng-ust1t64``
  * ``libbrotli1``
```

With the extra space, they're rendering as block quotes. rST is fiddly. :(

Author: So _that_ is how you fix this! Thanks! 😄

* ``launcher``: A part that sets up the snap launcher to run the .NET application. It verifies whether the necessary
  .NET runtime is available through a content interface connection and launches the application using the ``dotnet`` command.

.. collapse:: Included parts

    .. code-block:: yaml
        :caption: snapcraft.yaml

        launcher:
            plugin: dump
            source: /snap/snapcraft/x4/share/snapcraft/extensions/dotnet
            override-build: |
              mkdir -p $CRAFT_PART_INSTALL/bin/command-chain
              cp launcher.sh $CRAFT_PART_INSTALL/bin/command-chain
            stage:
            - bin/command-chain/launcher.sh
        prereqs:
            plugin: nil
            stage-packages:
            - libicu74
            - libunwind8
            - libssl3t64
            - liblttng-ust1t64

>             - libbrotli1

Author: It worked :)

Included plugs
--------------

The .NET extensions connect a snap-wide content plug to the .NET runtime content snap that matches the application's target
.NET version.

> .. collapse:: Included snap-wide runtime plug for the dotnet8 extension

Reviewer: This is a slight departure from what we normally do, but I'm warming up to it. Let's give it a try!

Author: You mean the collapsibles? I took inspiration from the other extensions' documentation, e.g., [Flutter](https://documentation.ubuntu.com/snapcraft/stable/reference/extensions/flutter-extension/) and [GNOME](https://documentation.ubuntu.com/snapcraft/stable/reference/extensions/gnome-extension/).
